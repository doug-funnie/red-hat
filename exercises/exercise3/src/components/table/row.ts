import glamorous, { CSSProperties } from 'glamorous';
import { rgba } from 'polished';
import { ThemedProps } from 'styles/theme';

type Props = ThemedProps<{
  isHeading?: boolean;
  onClick?(evt: React.MouseEvent<HTMLDivElement>): void;
}>;

const defaultProps: Partial<Props> = {
  isHeading: false,
};

export const Row = glamorous.div<Props>(({
  theme,
  isHeading,
  onClick,
}) => {
  const styles: CSSProperties = {
    alignItems: 'center',
    borderRadius: 4,
    display: 'flex',
    marginBottom: 5,
    transition: 'background-color 150ms ease-in-out',
    backgroundColor: !isHeading && theme.colors.white,
  };

  if (onClick) {
    return {
      cursor: onClick && 'pointer',
      ...styles,
      ':hover': {
        backgroundColor: rgba(theme.colors.red, 0.2),
      },
    };
  }
  return styles;
});

Row.defaultProps = defaultProps;
