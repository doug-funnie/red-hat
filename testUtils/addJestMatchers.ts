// import serializer from 'jest-glamor-react';
import { toMatchSnapshot } from 'jest-snapshot';

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeCalledWithSnapshot(snapshotName?: string): void;
    }
  }
}

// expect.addSnapshotSerializer(serializer);

expect.extend({
  toBeCalledWithSnapshot(received, snapshotName) {
    if (received.mock.calls.length === 0) {
      return {
        message: () => 'Mock function was not called',
        pass: false,
      };
    }

    return toMatchSnapshot.call(this, received.mock.calls, snapshotName);
  },
});
