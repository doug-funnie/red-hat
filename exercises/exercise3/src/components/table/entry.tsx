import * as React from 'react';
import { Cell } from './cell';
import { Column } from './column';
import { Row } from './row';

export type EntryClickHandler<D> = (props: { data: D, index: number, event: React.MouseEvent<HTMLDivElement> }) => void;

export interface Props<D = any> {
  data: D;
  index: number;
  columns: Column<D>[];
  onClick: EntryClickHandler<D>;
}

export class Entry extends React.Component<Props> {
  private handleRowClick = (event: React.MouseEvent<HTMLDivElement>) => {
   this.props.onClick({
      event,
      data: this.props.data,
      index: this.props.index,
    });
  }

  public render() {
    const {
      data,
      index,
      columns,
      onClick,
    } = this.props;

    return (
      <Row onClick={onClick ? this.handleRowClick : undefined}>
        {columns.map((column) => (
          <Cell key={column.id} grow={column.grow} hideOnMobile={column.hideOnMobile}>
            {column.renderCell({ data, index , column })}
          </Cell>
        ))}
      </Row>
    );
  }
}
