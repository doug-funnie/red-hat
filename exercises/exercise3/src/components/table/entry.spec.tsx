import { shallow } from 'enzyme';
import * as React from 'react';
import '../../../../../testUtils/addJestMatchers';
import { Entry, Props } from './entry';
import { Row } from './row';

const props: Props<number> = {
  data: 1,
  index: 0,
  onClick: jest.fn(),
  columns: [{
    id: 1,
    renderLabel: jest.fn(),
    renderCell: jest.fn(),
  }],
};

it('calls renderCell for column', () => {
  shallow(<Entry {...props} />);
  expect(props.columns[0].renderCell).toBeCalledWithSnapshot();
});

it('calls onClick on row click', () => {
  const view = shallow(<Entry {...props} />);
  view.find(Row).simulate('click', 'event');
  expect(props.onClick).toBeCalledWithSnapshot();
});
