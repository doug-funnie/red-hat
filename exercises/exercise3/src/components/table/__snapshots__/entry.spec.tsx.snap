// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls onClick on row click 1`] = `
Array [
  Array [
    Object {
      "data": 1,
      "event": "event",
      "index": 0,
    },
  ],
]
`;

exports[`calls renderCell for column 1`] = `
Array [
  Array [
    Object {
      "column": Object {
        "id": 1,
        "renderCell": [MockFunction] {
          "calls": [Circular],
        },
        "renderLabel": [MockFunction],
      },
      "data": 1,
      "index": 0,
    },
  ],
]
`;
